### ISSUE TITLE
[css-contain-3][css-mediaqueries-5] Custom container queries

### ISSUE THREAD

**Issue by Que-tin** [w3c/csswg-drafts#7622]
The spec for Media Queries Level 5 currently specifies a way to define custom media properties like e.g. `@custom-media --modern (color), (hover);` with CSS Containment currently getting shipped beforehand, wouldn't it make sense to rename @custom-media to @custom-query, to create a reusable and unified way for creating custom properties that are including a query?

I mean if I get the idea of the current draft correctly, values inside of  `@custom-media` could also be used inside of @container and @supports as well. But these aren't media, but feature & containment / container queries.

If from a technical stand point it isn't possible to get  a unified solution for queries, this is a proposal for a e.g. @custom-container for custom container queries.

**Comment by valtlai** [comment]
Related: #6698 (using custom `env()` variables in media queries)

**Comment by Que-tin** [comment]
Following recent discussions on this topic, I'd like to bring this topic back into focus

**Comment by andruud** [comment]
I have two questions surrounding `@custom-media`:
 
- Extrapolating from `@custom-media`, are we going to add `@custom-supports`, `@custom-container`, `@custom-whatever` in the future? (Is that nice?)
- Are we anyway going to add something else that makes `@custom-media` redundant? The css-env-1 spec already expects `env()` to work on MQs, and if we plan to do author-defined `env()`, then I'm not sure if `@custom-media` is needed at all. (EDIT: https://github.com/w3c/csswg-drafts/issues/6698)

**Comment by romainmenke** [comment]
I think a large part of the use case is simply to give a name to queries.
So the feature is not being able to reuse `768px` but instead to give a name to `(width >= 768px)`

Giving it a name implies several other aspects:
- listing all the named things gives an overview of all the predefined media queries in a project
- "jump to definition" and other niceties in code editors

I think this is mostly useful for `@media` and `@container`.
I am not convinced about `@supports` needing this for the same reasons.

This isn't that convenient with `env()` because there is nothing about a custom environment variable that communicates its intended use. You can't extrapolate "this project has a breakpoint on `768px`" from a custom env variable, it might be defined for unrelated reasons.

**Comment by Que-tin** [comment]
@romainmenke with [CSS Conditional Rules Module Level 5](https://www.w3.org/TR/css-conditional-5/#when-rule) defining `@when` as well as `@else` that accepts both `media()` and `supports()` as well as in combination, I think leaving `@supports` out of the loop when thinking about this is probably not the way to go.

But I'm with you that `env()` is more thought of as a way to save globally available values instead of full conditional rules.

`env()` should be used in combination with `@custom-media` | `@custom-query` , but shouldn't replace it.

**Comment by andruud** [comment]
@romainmenke Yeah, I do see that. Though those benefits are not such much about the thing having a _name_, but more about the thing having a _type_.

So you think `@custom-<anything>` is indeed the way to go, then?

**Comment by romainmenke** [comment]
The thing having a type is indeed important.

If `@custom-<anything>` is implementable and isn't an excessive burden to implementors then, yes I do think it is the way to go :)

But I don't have any strong attachment to the currently specified syntax.
Something more general purpose would also be fine.

Maybe something inspired by `@property`?:

```css
@prelude --foo {
  syntax: "@media"; /* with some specific syntax to reference "whatever is valid in that other prelude" */
  value: (width >= 768px);
}
```

**Comment by andruud** [comment]
> If `@custom-<anything>` is implementable and isn't an excessive burden to implementors then, yes I do think it is the way to go :)

It's quite implementable, and not a burden.

**Comment by romainmenke** [comment]
I also wrote some tests for `@custom-media` some time ago as part of the interop proposal: https://github.com/web-platform-tests/wpt/pull/48480 Might be useful

**Comment by jimmyfrasche** [comment]
One thing in custom media specifically was the ability to use JS to override preference queries so that, for example, you could set dark mode based on a toggle on the page but just use regular media queries in the css. That may need to be moved elsewhere.

**Comment by romainmenke** [comment]
@jimmyfrasche do you have a link?

Current specification only has this: https://drafts.csswg.org/mediaqueries-5/#script-custom-mq

>  Define a map of names to values for JS. Values can be either a MediaQueryList object or a boolean, in which case it’s treated identically to the above, or can be a number or a string, in which case it’s treated like a normal MQ, and can use the normal or range context syntax. Like:
> 
> ```html
> <script>
> CSS.customMedia.set('--foo', 5);
> </script>
> <style>
> @media (_foo: 5) { ... }
> @media (_foo < 10) { ... }
> </style>
> ```

**Comment by jimmyfrasche** [comment]
It's in https://drafts.csswg.org/mediaqueries-5/#auto-pref%E2%91%A0 but it looks like it has its own API now (last time I checked it was using the `CSS.customMedia.set()` function)

**Comment by brandonmcconnell** [comment]
Initially, I was a bigger fan of the more arbitrary `@custom-query` name over `@custom-media`, as it provided unified support across all query types. However, the more I thought about it, size names typically hold different meanings in different contexts. `@container (--small)` and `@media (--small)` may mean two very different things — small screens vs. small elements.

For that reason, I think something like @romainmenke's suggestion of an at-rule rule like `@prelude` for defining these could make a lot of sense, as an alternative to separate `@custom-media`, `@custom-container`, and `@custom-supports` queries.

One useful addition I'd propose would be the option to explicitly support multiple query types:

Something like this:
```postcss
/* only media */
@query-value --foo {
  syntax: media;
  value: (width >= 768px);
}

/* media and container */
@query-value --foo {
  syntax: media, container;
  value: (width >= 768px);
}
```

I also think it might be better to use query names by their idents _without_ `@`, since those are their names, which can also be used via `media()`, `container()`, etc. when used in `@if` queries— which we should also consider here.

**Comment by brandonmcconnell** [comment]
Working through all of that, would `env()` be the most flexible solution, after all. What are the pros and cons/gotchas of each that we would need to consider?

**Comment by andruud** [comment]
> Maybe something inspired by `@property`?:
> ```
> @prelude --foo {
>   syntax: "@media"; /* with some specific syntax to reference "whatever is valid in that other prelude" */
>   value: (width >= 768px);
> }
> ```

Perhaps we don't need to be flexible "at any cost" (in this case: verbosity/readability). 

Also: if we go this route, we'll probably want to avoid a syntax that can be a number of different things (e.g. `syntax: media, container`), because tooling may not be able to know what we're looking at anymore.

**Comment by tabatkins** [comment]
I don't think I see the reason to do a "generic" at-rule that still has to specify what type of condition it's for. At that point you're just shifting the specifier around, and it's more ergonomic to put it in the name of the at-rule instead.

That is, `@prelude --foo { syntax: "@media"; value: (width >= 768px); }` is identical to `@custom-media --foo (width >= 786px);`, but the latter is shorter and gives us a better ability to expose a half-useful CSSOM class (and better tooling in general).

And I think I agree with Anders that allowing a single rule to specify a value for multiple types of rules isn't great, since their syntaxes only overlap by accident.

I'm okay, I think, with a proliferation of `@custom-<condition>` rules; we don't have too many of them *anyway*, and it makes things clearer overall.

**Comment by kbabbitt** [comment]
For combining conditions of different types, what about something like:

```css
@prelude --foo {
  @media (width >=768px);
  @supports (anchor-name: --blah);
}
```

Or perhaps something similar to the grammar for `@when`:

```css
@prelude --foo (media(width >= 768px) and supports(anchor-name: --blah));
```

Having said that, I do also like the compactness and clarity of e.g. `@custom-media --foo (width >=768px);` when multiple types of conditions aren't needed.

**Comment by tabatkins** [comment]
But like, where would that `--foo` "prelude" be used? It looks like it's only valid in `@when`. We can just have a `@custom-when` for that, right?

Or is the thought that it would function as an MQ in `@media`, a supports query in `@supports`, and both at once in `@when`?

**Comment by kbabbitt** [comment]
I was imagining it being valid only for `@when` ... yeah, I suppose it would just be `@custom-when`.
